<a href="https://colab.research.google.com/github/muhammadhafiz27/Books-To-Scrape/blob/main/Colab/BooksToScrape.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Import Data**

In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/muhammadhafiz27/Books-To-Scrape/refs/heads/main/books_to_scrape_tugas.csv"
df = pd.read_csv(url)
display(df.head())

web-scraper-order                             web-scraper-start-url  \
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   

                                               Judul  \
0  In Her Wake\n\nHow Music Works\n\nFoolproof Pr...   
1  A Light in the ...\n\nTipping the Velvet\n\nSo...   

                                               harga  \
0  £12.84\n\n£37.32\n\n£30.52\n\n£25.27\n\n£34.53...   
1  £51.77\n\n£53.74\n\n£50.10\n\n£47.82\n\n£54.23...   

                                              rating  
0  star-rating One\nstar-rating Two\nstar-rating ...  
1  star-rating Three\nstar-rating One\nstar-ratin...

In [5]:
df_cleaned = df.copy()
for col in ['Judul', 'harga']:
    df_cleaned[col] = df_cleaned[col].str.split('\n\n')
df_cleaned['rating'] = df_cleaned['rating'].str.split('\n')
df_cleaned = df_cleaned.explode(['Judul', 'harga', 'rating'])
display(df_cleaned.head(40))

web-scraper-order                             web-scraper-start-url  \
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   

                                   Judul   harga             rating  
0                            In Her Wake  £12.84    star-rating One  
0                        How Music Works  £37.32    star-rating Two  
0      Foolproof Preserving: A Guide ...  £30.52  star-rating Three  
0             Chase Me (Paris Nights ...  £25.27   star-rating Five  
0                             Black Dust  £34.53   star-rating Five  
0               Birdsong: A Story in ...  £54.64  star-rating Three  
0  America's Cradle of Quarterbacks: ...  £22.50  star-rating Three  
0          Aladdin and His Wonderful ...  £53.13  star-rating Three  
0  Worlds Elsewhere: Journeys Around ...  £40.30   star-rating Five  
0                         Wall and Piece  £44.18   star-rating Four  
0             The Four Agreements: A ...  £17.66   star-rating Five  
0           The Five Love Languages: ...  £31.05  star-rating Three  
0                      The Elephant Tree  £23.82   star-rating Five  
0                   The Bear

# **Analisis Sederhana**

**Rata-Rata Harga Buku**

In [6]:
df_cleaned['harga_numeric'] = df_cleaned['harga'].str.replace('£', '').astype(float)
average_price = df_cleaned['harga_numeric'].mean()
print(f"The average price of the books is £{average_price:.2f}")

The average price of the books is £34.96


**Buku dengan Rating “Five”**

In [8]:
five_star_books = df_cleaned[df_cleaned['rating'] == 'star-rating Five']
num_five_star_books = len(five_star_books)
print(f"There are {num_five_star_books} books with a 'Five' rating.")

There are 10 books with a 'Five' rating.


# **Task**
Analisis data buku dari "https://raw.githubusercontent.com/muhammadhafiz27/Books-To-Scrape/refs/heads/main/books_to_scrape_tugas.csv" untuk menghitung rata-rata harga buku dan jumlah buku dengan rating "Five", lalu berikan langkah-langkah pengerjaan dalam bahasa Indonesia.

## Muat data

### Subtask:
Muat data buku dari URL yang disediakan ke dalam pandas DataFrame.


**Reasoning**:
Load the data into a pandas DataFrame and display the head.



In [11]:
url = "https://raw.githubusercontent.com/muhammadhafiz27/Books-To-Scrape/refs/heads/main/books_to_scrape_tugas.csv"
df = pd.read_csv(url)
display(df.head())

web-scraper-order                             web-scraper-start-url  \
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   

                                               Judul  \
0  In Her Wake\n\nHow Music Works\n\nFoolproof Pr...   
1  A Light in the ...\n\nTipping the Velvet\n\nSo...   

                                               harga  \
0  £12.84\n\n£37.32\n\n£30.52\n\n£25.27\n\n£34.53...   
1  £51.77\n\n£53.74\n\n£50.10\n\n£47.82\n\n£54.23...   

                                              rating  
0  star-rating One\nstar-rating Two\nstar-rating ...  
1  star-rating Three\nstar-rating One\nstar-ratin...

## Bersihkan data

### Subtask:
Proses data yang dimuat dengan memecah dan meledakkan kolom 'Judul', 'harga', dan 'rating' untuk memastikan setiap baris mewakili satu buku dengan detail yang sesuai.


**Reasoning**:
Process the loaded data by splitting and exploding the 'Judul', 'harga', and 'rating' columns to ensure each row represents a single book with its corresponding details.



In [12]:
df_cleaned = df.copy()

for col in ['Judul', 'harga']:
    df_cleaned[col] = df_cleaned[col].str.split('\n\n')

df_cleaned['rating'] = df_cleaned['rating'].str.split('\n')

df_cleaned = df_cleaned.explode(['Judul', 'harga', 'rating'])

display(df_cleaned.head(40))

web-scraper-order                             web-scraper-start-url  \
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
0      1757757531-1  https://books.toscrape.com/catalogue/page-2.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   
1      1757757533-2  https://books.toscrape.com/catalogue/page-1.html   

                                   Judul   harga             rating  
0                            In Her Wake  £12.84    star-rating One  
0                        How Music Works  £37.32    star-rating Two  
0      Foolproof Preserving: A Guide ...  £30.52  star-rating Three  
0             Chase Me (Paris Nights ...  £25.27   star-rating Five  
0                             Black Dust  £34.53   star-rating Five  
0               Birdsong: A Story in ...  £54.64  star-rating Three  
0  America's Cradle of Quarterbacks: ...  £22.50  star-rating Three  
0          Aladdin and His Wonderful ...  £53.13  star-rating Three  
0  Worlds Elsewhere: Journeys Around ...  £40.30   star-rating Five  
0                         Wall and Piece  £44.18   star-rating Four  
0             The Four Agreements: A ...  £17.66   star-rating Five  
0           The Five Love Languages: ...  £31.05  star-rating Three  
0                      The Elephant Tree  £23.82   star-rating Five  
0                   The Bear

## Hitung Rata-Rata Harga Buku

### Subtask:
Hitung rata-rata harga buku dari kolom 'harga' yang telah dibersihkan.

**Reasoning**:
To calculate the average price, I need to convert the 'harga' column to a numeric format by removing the currency symbol and converting the type. Then I can calculate the average.

In [13]:
df_cleaned['harga_numeric'] = df_cleaned['harga'].str.replace('£', '').astype(float)

average_price = df_cleaned['harga_numeric'].mean()

print(f"The average price of the books is £{average_price:.2f}")

The average price of the books is £34.96


## Hitung Buku dengan Rating “Five”

### Subtask:
Hitung jumlah buku dengan rating "Five" dari kolom 'rating' yang telah dibersihkan.

**Reasoning**:
To find the number of books with a "Five" rating, I will filter the DataFrame to select only those rows where the rating is 'star-rating Five' and then count the number of rows.

In [14]:
five_star_books = df_cleaned[df_cleaned['rating'] == 'star-rating Five']

num_five_star_books = len(five_star_books)

print(f"There are {num_five_star_books} books with a 'Five' rating.")

There are 10 books with a 'Five' rating.


## Ringkasan Analisis

### Subtask:
Rangkum temuan dari analisis rata-rata harga dan jumlah buku dengan rating "Five".

**Reasoning**:
Summarize the key findings from the analysis.

### Temuan Utama Analisis Data

* **Rata-Rata Harga Buku**: Harga rata-rata buku adalah £34.96.
* **Jumlah Buku Rating "Five"**: Terdapat 10 buku dengan rating "Five".

Tugas analisis data buku telah selesai dengan menghitung rata-rata harga buku dan jumlah buku dengan rating "Five".